### PR TITLE
Add Xgc options for suballocator heap size and quick allocation

### PR DESCRIPTION
**Pull request by ThanHenderson**
1. Use VMEM_ALLOC_QUICK by default for allocateRegion in allocate_memory32
2. Adds -Xgc:suballocatorQuickAllocDisable option that disables the default VMEM_ALLOC_QUICK
3. Adds -Xgc:suballocatorIncrementSize option that replaces the HEAP_SIZE_BYTES macro and controls the heap increment size
4. Adds sanity.functional tests

Addresses: https://github.com/eclipse/omr/issues/7190
Signed-off-by: Nathan Henderson <nathan.henderson@ibm.com>

Depends on: https://github.com/eclipse/omr/pull/7472

### CODE REVIEW

**Comment by ThanHenderson** [comment]
@babsingh this is the accompanying OpenJ9 patch for https://github.com/eclipse/omr/pull/7414 that needs to be coordinated with it.

File: runtime/gc_modron_startup/mmparseXgc.cpp
Comment (dmitripivkine) [inline runtime/gc_modron_startup/mmparseXgc.cpp:1464]: This condition should be exactly the same as in OMR MM_GCExtensionsBase" `if defined(OMR_ENV_DATA64) && defined(LINUX)`. Alternatively (and this is my preference) this field mighty exist for all platforms (despite is used on Linux only), at least in GC part of the code.

Comment (ThanHenderson) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1464]: Sounds good. I will just remove the directives from this patch and the omr patch.

File: runtime/oti/j9port.h
Comment (keithc-ca) [inline runtime/oti/j9port.h:783]: I don't think we should be adding to this list of macros: just use the ones provided by OMR.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml:22]: Please remove the indentation of these lines.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml:30]: Where is the restriction to compressedrefs suggested by the comment?

File: test/functional/cmdLineTests/gcsuballoctests/playlist.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/playlist.xml:58]: This should also specify that a 64-bit VM is required.

File: runtime/oti/j9port.h
Comment (ThanHenderson) [reply runtime/oti/j9port.h:783]: The macro defined on the line above, i.e.:
```
#define J9PORT_CTLDATA_ALLOCATE32_INCREMENT_SIZE OMRPORT_CTLDATA_ALLOCATE32_INCREMENT_SIZE
```
is also not used. Should that be removed too?

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml
Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballocquickalloctest.xml:22]: I will be removing this file anyways since we are no longer using the directives specifying that `suballocatorQuickAllocDisable` is on Linux only.

File: runtime/gc_modron_startup/mmparseXgc.cpp
Comment (dmitripivkine) [inline runtime/gc_modron_startup/mmparseXgc.cpp:1469]: Please add `suballocatorQuickAllocEnable` option as well for symmetry.

Comment (ThanHenderson) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1469]: I've added it. But in the OMR patch: https://github.com/eclipse/omr/blob/6a68a292908d30fb6552a46d585d87af33f21904/port/common/omrportcontrol.c#L93-L103 
I've left handling for this case as a no-op since `true` is the default for `suballocatorQuickAlloc`. Let me know if you think I should add in case in `omrport_control` anyways.

Comment (dmitripivkine) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1469]: `omrport_control(OMRPORT_CTLDATA_ALLOCATE32_QUICK_ALLOC, extensions->suballocatorQuickAlloc);` call from Memory Manager should set key value regardless it's default value. What prevents just to do `PPG_mem_mem32_subAllocHeapMem32.suballocator_quickAlloc = value;`?

Comment (ThanHenderson) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1469]: Nothing prevents it. I was just trying to follow similar logic to what was provided with the other options.

I can change it to that.

Comment (dmitripivkine) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1469]: May be (I am not insisting) because of `value` parameter has `uintptr_t` type it is better to use this type explicitly at both sides instead of `bool`/`BOOLEAN`? In GC for  `extensions->suballocatorQuickAlloc` and in Port for `PPG_mem_mem32_subAllocHeapMem32.suballocator_quickAlloc`, just to prevent types conversion?

Comment (ThanHenderson) [reply runtime/gc_modron_startup/mmparseXgc.cpp:1469]: I like that idea too.

**Review by dmitripivkine — APPROVED** [review]
Looks good to me

File: runtime/gc_modron_startup/mmparseXgc.cpp
Comment (keithc-ca) [inline runtime/gc_modron_startup/mmparseXgc.cpp:1450]: Missing "for" (or similar) in "request for a".

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml:31]: I don't see anything in the code that limits the options to compressedrefs mode: Please explain.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml:54]: The test id should match reality: 16m vs. 100m.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml:126]: This comment is stale: the library is `j9gc29`; could you please fix this and the others?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml:31]: By "the code" do you mean in places like [this](https://github.com/eclipse/omr/blob/576481c1ac4e34da29886a444d19574bc029156e/gc/base/GCExtensionsBase.hpp#L747) not being behind `#if defined(OMR_GC_COMPRESSED_POINTERS)` guards?

The `suballocator...` options are only ever used by code guarded by `#if defined(OMR_GC_COMPRESSED_POINTERS)`. I assumed the previous `suballoctor...` command-line options were still defined for non-compressed refs builds for Java command invocation compatibility between non-compressed refs builds and compressed refs builds, so I followed that decision.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [inline test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml:28]: While you're here, please remove the trailing whitespace here and on line 95 and remove the extra blank line at the end of this file.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: The options are recognized unconditionally (there's no `#if defined(OMR_GC_COMPRESSED_POINTERS)` guarding that code in `mmparseXgc.cpp`), but I see the use of `suballocatorIncrementSize` in eclipse/omr#7414 is guarded - perhaps the field declaration should be as well?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: Yes, that was the point I was trying to make in my previous comment.

I think it is better if the options aren't guarded " for Java command invocation compatibility between non-compressed refs builds and compressed refs builds" like I mentioned.

Is your opinion that we guard them? And also @dmitripivkine what is yours?

Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: I wasn't suggesting that the option parsing be conditional, just the presence of the fields.

Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: Right. So then in the `playlist.xml` file for this test, do you know how I could specify that it should only be run on compressed reference builds?

Comment (dmitripivkine) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: I would prefer not to wrap field as well as option recognition in this case. The motivation is we prefer to recognize the option always to allow customers to use command line flexibly. And guarding inside option recognition does not make much sense, so my vote is not use wrapping for field declaration too.

Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: If we agree with Dmitri, the most recent push addresses all of your comments @keithc-ca

Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: It's not my preference to have fields that are not used. The more serious problem is using when they should not be, but I can live with the fields being present even if they're not used.

Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: > how I could specify that it should only be run on compressed reference builds?

Perhaps @llxia can help with that.

Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: That is only necessary if we `#if defined(OMR_GC_COMPRESSED_POINTERS)` guard the parsing. Otherwise, we can just leave it as it is.

Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: No. The possibility is that someone adds uses of the fields thinking they're relevant when they're not. If the field were conditional, those uses would have to be conditional as well. We should let the compiler help us with things like that, but as I said, I won't insist on making the fields conditional (which would be an OMR committer's decision).

Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: I was referring to changing the tests to run only for compressed refs. I agree with you on the point you just made.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: I still see a trailing space on line 95.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (keithc-ca) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: Looking at the tests again, I think you're right - they seem mostly about option validation which happens independently of whether compressed refs are in use or not.

File: test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml
Comment (ThanHenderson) [reply test/functional/cmdLineTests/gcsuballoctests/gcsuballoctests.xml]: Should be good now.

**Comment by keithc-ca** [comment]
> Needs to be coordinated with: [eclipse/omr#7414](https://github.com/eclipse/omr/pull/7414)

I think the OMR change can safely be merged without this, or did I miss something?

**Review by keithc-ca — APPROVED** [review]

**Comment by ThanHenderson** [comment]
The OMR patch has been merged. We can start the PR builds for this one too.

**Comment by dmitripivkine** [comment]
-Xgc: options are documented. Would you please create doc item to add description for new options?

**Comment by pshipton** [comment]
The OMR changes also need to promote, but PR builds can depend on OMR #master.
Currently promotion is blocked. When https://github.com/eclipse/omr/pull/7430 is merged that might unblock it, or we'll see what the next problem is.

File: runtime/exelib/common/memcheck.c
Comment (keithc-ca) [inline runtime/exelib/common/memcheck.c:758]: Why shouldn't this go in `memoryCheck_initialize()` instead?

Comment (ThanHenderson) [reply runtime/exelib/common/memcheck.c:758]: You're right. I was just thinking this too. It doesn't need to be called every time we allocate.

Comment (ThanHenderson) [reply runtime/exelib/common/memcheck.c:758]: Actually. I remember why I did it this way. The original `portLib` `PPG` options are not yet set when `memCheck_initialize` is called; that is why I pushed it back to the allocator wrapper.

Comment (keithc-ca) [reply runtime/exelib/common/memcheck.c:758]: I still don't think here is the right place to propagate those options. It should happen just once in the life-cycle of memcheck.

Comment (ThanHenderson) [reply runtime/exelib/common/memcheck.c:758]: I agree. I'm looking into exactly where a more appropriate placement would be.

Comment (ThanHenderson) [reply runtime/exelib/common/memcheck.c:758]: If we are to change where we set up the `PPG` options for the `memCheckPortLib`, the call to the set up routine would need to be [here](https://github.com/eclipse-openj9/openj9/blob/028a24e1522017110a37a883cd9c1de10f9c7291/runtime/vm/jvminit.c#L7512).

The preceding condition calls `runInitializationStage(vm, ALL_LIBRARIES_LOADED)` which sets up the `PPG` options in the source `portLib`. Within the call in the next line, `J9RelocateRASData(vm);`, there is an allocation through the suballocator. When `-Xcheck` is used, it uses the `memCheckPortLib` sub-allocator wrapper.

We could add a function pointer on the `portLib` that would be assigned in `memoryCheck_initialize` to a procedure defined in `memcheck.c` that calls `port_copy_suballocator_heap_globals`. Then we'd call that function in the location where I've linked to above.

I'd prefer to leave things as they are now (since the setup is contained within the `memcheck.c`), instead of an approach similar to the one that I've described.

@keithc-ca let me know what you think, or if you had any other ideas.

Comment (ThanHenderson) [reply runtime/exelib/common/memcheck.c:758]: @keithc-ca I'd like to get this merged. Could you give some input on what I have said above?

**Comment by ThanHenderson** [comment]
I've removed the ability to propagate the port globals from one port lib to another; we can just rely on default values now specified on the OMR side to ensure that the `memoryCheckPortLib` has some non-zero value for the `suballocatorIncrementSize`. We no longer need a coordinated merge. `memoryCheckPortLib` just will not get the values from the `-Xgc` options. We can revisit that in the future, and I think the memcheck support should be updated anyways to make it more amenable to proper initialization.

fyi @babsingh

**Comment by babsingh** [comment]
jenkins line endings check

**Comment by pshipton** [comment]
Seems like we should have a doc issue opened for this.
https://github.com/eclipse-openj9/openj9-docs/issues/

**Review by babsingh — APPROVED** [review]
These changes LGTM. In the PR builds,

I saw three known issues:
- https://github.com/eclipse-openj9/openj9/issues/13223
- https://github.com/eclipse-openj9/openj9/issues/13684
- https://github.com/eclipse-openj9/openj9/issues/18463

Two undocumented issues were seen:
- https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.openjdk_x86-64_windows_Personal_testList_2/26: `PrimeTest` segfaulted while running a JIT'ed method, `PrimeTest.createPrimes`.
- https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.openjdk_s390x_linux_Personal_testList_0/28: `MicroTime` failed with `Exception: What? only 138 musec precision?`

None of the above failures occur due to this PR's changes. @ThanHenderson Can you please open new issues to document the above two failures (`PrimeTest` & `MicroTime`)?

I will merge this PR once the latest OMR acceptance build promotes https://github.com/eclipse/omr/pull/7472: https://openj9-jenkins.osuosl.org/job/Pipeline-OMR-Acceptance.

**Comment by babsingh** [comment]
Merging. https://github.com/eclipse/omr/pull/7472 has promoted, and the corresponding changes show in https://github.com/eclipse-openj9/openj9-omr/commits/openj9.

**Pull request merged**
